### PR TITLE
[autoparallel] resnet block runtime apply

### PR DESCRIPTION
**Pull request by YuliangLiu0306**
### What does this PR do
1. Previously, we do not consider the module buffers such as running mean and running variance. In this PR, **the module buffers is considered during strategy generating and runtime applying.**
2. **fix a bug in CostGraph** which leads to successors of node removed accidentally.
3. Build a runtime unit test for resnet Bottleneck Block.

### CODE REVIEW

File: tests/test_auto_parallel/test_resnet_block_runtime.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    input = torch.rand(256, 64, 64, 64).cuda()
+    physical_mesh_id = torch.arange(0, 4)
+    mesh_shape = (2, 2)
+    # [[0, 1]
+    #  [2, 3]]
+    device_mesh = DeviceMesh(physical_mesh_id, mesh_shape, init_process_group=False)
+    entire_shape = torch.Size((4, 4, 8, 8))
+
+    tracer = ColoTracer()
+    model = Bottleneck(64, 64, 1, norm_layer=torch.nn.modules.batchnorm.BatchNorm2d).cuda()
```
Comment (FrankLeeeee) [inline tests/test_auto_parallel/test_resnet_block_runtime.py:119]: You should be able to import this module from `torchvision` directly.

File: tests/test_auto_parallel/test_resnet_block_runtime.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    gm.recompile()
+    nodes = [node for node in gm.graph.nodes]
+    # TODO: wrap the gm to avoid the influence of the user training code
+    torch.cuda.set_rng_state(cuda_rng_state)
+    output = gm(input, sharding_spec_dict, origin_spec_dict)
+    assert output.shape == origin_output.shape
+    assert output.equal(origin_output)
+
+
+@pytest.mark.dist
```
Comment (FrankLeeeee) [inline tests/test_auto_parallel/test_resnet_block_runtime.py:161]: Mark this with `run_on_env_flag` if it takes a while.

File: colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py [hunk 3]
```diff
@@ -274,15 +306,15 @@ def generate(self):
...
 
         # SS = SS x S WITH SYNC_BN
-        strategy_list.append(self.split_input_both_dim(0, 1))
-        strategy_list.append(self.split_input_both_dim(1, 0))
+        # strategy_list.append(self.split_input_both_dim(0, 1))
+        # strategy_list.append(self.split_input_both_dim(1, 0))
 
         # S01R = S01R x R WITH SYNC_BN
-        strategy_list.append(self.split_input_batch_1d(0, 1))
+        # strategy_list.append(self.split_input_batch_1d(0, 1))
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py:317]: Why comment out these strategies?

File: colossalai/device/device_mesh.py [hunk 4]
```diff
@@ -52,6 +52,14 @@ def __init__(self,
         if self.need_flatten:
             self.flatten_device_mesh = self.flatten()
 
+    # def __deepcopy__(self, memo):
+    #     obj = DeviceMesh(self.physical_mesh_id, self.mesh_shape, self.mesh_alpha, self.mesh_beta, init_process_group=False, need_flatten=False)
+    #     if hasattr(self, 'process_groups_dict'):
+    #         obj.process_groups_dict = self.process_groups_dict
+    #     if hasattr(self, 'flatten_device_mesh'):
+    #         obj.flatten_device_mesh = self.flatten_device_mesh
+    #     return obj
```
Comment (FrankLeeeee) [inline colossalai/device/device_mesh.py:61]: Why leave them commented out?

File: colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py [hunk 5]
```diff
@@ -76,23 +78,23 @@ def update_memory_cost(self, strategy: ShardingStrategy):
         # compute fwd cost incurred
         # fwd_cost = input + other + bias + output
         fwd_activation_cost = sum([v for k, v in forward_size_mapping.items() if not self.is_param(k)])
-        fwd_parameter_cost = sum([v for k, v in forward_size_mapping.items() if self.is_param(k)])
-        fwd_mem_cost = MemoryCost(activation=fwd_activation_cost, parameter=fwd_parameter_cost)
+        fwd_parameter_and_buffer_cost = sum([v for k, v in forward_size_mapping.items() if self.is_param(k)])
+        fwd_mem_cost = MemoryCost(activation=fwd_activation_cost, parameter=fwd_parameter_and_buffer_cost)
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py:82]: Since we differentiate parameter and buffer in the `OperationDataType`, we should separate them in the `MemoryCost` as well.

File: tests/test_auto_parallel/test_resnet_block_runtime.py (hunk 1 above)
Comment (FrankLeeeee) [reply tests/test_auto_parallel/test_resnet_block_runtime.py:119]: It can be found here.
https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py#L108

Comment (YuliangLiu0306) [reply tests/test_auto_parallel/test_resnet_block_runtime.py:119]: This code is copied from torchvision, the residual add is removed in this PR, because broadcast op handler is not ready.

File: colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py (hunk 3 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py:317]: We need some additional pass to use SyncBN substitute BatchNorm, it will be done in future PR.

File: colossalai/device/device_mesh.py (hunk 4 above)
Comment (YuliangLiu0306) [reply colossalai/device/device_mesh.py:61]: I will delete it.

File: colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py (hunk 3 above)
Comment (FrankLeeeee) [reply colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py:317]: OK, perhaps some TO-DOs and comments for this part of code will help us understand better.

File: tests/test_auto_parallel/test_resnet_block_runtime.py (hunk 1 above)
Comment (FrankLeeeee) [reply tests/test_auto_parallel/test_resnet_block_runtime.py:119]: Ok, understand.

File: colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py (hunk 5 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py:82]: Okay.

File: tests/test_auto_parallel/test_resnet_block_runtime.py (hunk 2 above)
Comment (YuliangLiu0306) [reply tests/test_auto_parallel/test_resnet_block_runtime.py:161]: Done

Comment (FrankLeeeee) [reply tests/test_auto_parallel/test_resnet_block_runtime.py:161]: I don't see it here?

Comment (YuliangLiu0306) [reply tests/test_auto_parallel/test_resnet_block_runtime.py:163]: Sorry, forget to push it.

File: colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py (hunk 3 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/tensor_shard/node_handler/strategy/batch_norm_generator.py:320]: Sure, I will add TODOs  and comments to explain the reason those strategies are commented.

**Review by FrankLeeeee — APPROVED** [review]
